### DISCUSSION TITLE
XML Schema for Geany-LSP

### DISCUSSION THREAD

**Discussion by Cicorione** [techee/geany-lsp#41] [Q&A] [answered]
Hi there,

I have been looking forward for this plugin for years!

I am not a coder, I am a designer that use Speedata Publisher to do create digital and offset printing documents. This LuaTeX based typography software uses XML schema to organize its xml structure. So far, as XML editor + schema, I have been using VSCode with Scholarly XML, but using Geany would be for me a great achievement for many reasons.

`Scholarly XML`:  https://github.com/raffazizzi/vscode-sxml ;  doesn't use java, I don't know if these VSCode plugins are pluggable from a software to another, but maybe it might be ported. Unfortunately I do not have the skills nor the knowledge to do so.

Alternatively is available `Lemminix`, which is the underline technology behind another VSCode plugins called `XML` (by Red Hat) and it is distribute as binary version that doesn't require java as well : https://github.com/eclipse/lemminx; current download is here: https://download.jboss.org/jbosstools/vscode/stable/lemminx-binary/

I don't know if these XML Language Server are compatible with `Geany-LSP` and worst I don't know how to setup theme. If you can help me, I am willing to report back anything and contribute to make this plugins better and better.

Thanks in advance! 🙏

Cicorione!

**Comment by techee** [comment]
> Scholarly XML: https://github.com/raffazizzi/vscode-sxml ; doesn't use java, I don't know if these VSCode plugins are pluggable from a software to another, but maybe it might be ported. Unfortunately I do not have the skills nor the knowledge to do so.

It doesn't really matter in what language the LSP server is written (it's run as a separate process independently of the editor) - it just has to be a LSP server. `vscode-sxml` seems to be a plugin for vscode and not a LSP server so it won't work.

> Alternatively is available Lemminix, which is the underline technology behind another VSCode plugins called XML (by Red Hat) and it is distribute as binary version that doesn't require java as well : https://github.com/eclipse/lemminx; current download is here: https://download.jboss.org/jbosstools/vscode/stable/lemminx-binary/

This one seems to be a LSP server. I'm on an ARM macBook with a ARM linux virtual machine so the x86 binaries won't work for me but you can give it a try. In principle just follow

https://github.com/techee/geany-lsp?tab=readme-ov-file#quick-start

and in step 3 add something like this to the configuration file
```ini
[XML]
cmd=/full/path/to/the/downloaded/binary/lemminx-linux
```
Also check the documentation of `lemminx` if it requires some extra command-line parameters which you can add as well.

**Reply by Cicorione** [reply]
Thank you, we are almost on the same page. I was interested to run it on aarch64... Now I realized that there aren't binary for aarch64, nor for FreeBSD even though I could use the linux emulation to have that server up & running... 🤔

**Reply by techee** [reply]
Note that for now the process of the LSP server has to be started by the plugin and communicate with Geany using stdin/stdout - sockets and remote connection isn't supported yet. So when you start a VM with the server inside the OS where you run Geany, you won't be able to connect the server.

**Comment by techee** [comment]
All right, I got an email notification with your post but don't see it anywhere here, strange.

You mention you tried
```ini
[XML]
cmd="java -jar /home/cicorione/.vscode/extensions/redhat.vscode-xml-0.27.1/server/org.eclipse.lemminx-0.28.0-uber.jar"
```
and got this error
```
LSP server process "java -jar /home/cicorione/.vscode/extensions/redhat.vscode-xml-0.27.1/server/org.eclipse.lemminx-0.28.0-uber.jar" failed to start, giving up: Failed to execute child process “"java” (No such file or directory)
```
The problem is that there should be no apostrophes after `=`. Notice the message `“"java” (No such file or directory)` and the 2 apostrophes there - this basically means the server tries to execute `"java` which it cannot find.

Try
```
[XML]
cmd=java -jar /home/cicorione/.vscode/extensions/redhat.vscode-xml-0.27.1/server/org.eclipse.lemminx-0.28.0-uber.jar
```

Also, try this command first in the console - it should just block until you terminate it with Ctrl+C. If it just prints some info and terminates by itself, have a look at its documentation and make sure the server is configured so it communicates with the client over stdin/stdout.

**Reply by Cicorione** [reply]
Sorry, I deleted because I thought was a pebcak and I decided to investigate more...

**Reply by Cicorione** [reply]
Hi @techee 

Without the `" "` the server is up and running, unfortunately I haven't found a way to validate the `xml` yet.

The server I am using is [Lemminx](https://github.com/eclipse/lemminx) that supports `xsd` as schema validation, but none of the below methods worked out based on these recommendations:
1. https://github.com/redhat-developer/vscode-xml/blob/main/docs/Validation.md
2. https://www.w3.org/TR/xml-model/

#### Method 1
```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="/home/cicorione/git/publisher/schema/layoutschema-en.xsd" type="application/xml" schematypens="http://http://www.w3.org/2001/XMLSchema"?>
```

#### Method 2
```xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
	xmlns:sd="urn:speedata:2009/publisher/functions/en"
	xmlns:xi="http://www.w3.org/2001/XInclude"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="/home/cicorione/git/publisher/schema/layoutschema-en.xsd">
```

#### Methode 3
```xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
	xmlns:sd="urn:speedata:2009/publisher/functions/en"
	xmlns:xi="http://www.w3.org/2001/XInclude"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="/home/cicorione/git/publisher/schema/layoutschema-en.xsd /home/cicorione/git/publisher/schema/layoutschema-en.xsd">
```

I wonder if it may be possible specify the schema in configuration file, e.g.:
```sh
[XML]
cmd=java -jar /home/cicorione/.vscode/extensions/redhat.vscode-xml-0.27.1/server/org.eclipse.lemminx-0.28.0-uber.jar
schema=/the/path/of/the/schema.xsd
```

Besides this  issue ,which is critical, I couldn't find any info on the taskbar that say anything about the status of the XML, so far all the editors I have used (VSCode & Emacs) tell you if the schema is associated or missing, and when the schema is associated it says if your xml is valid or invalid:

![vscode-bar](https://github.com/user-attachments/assets/91c4c0ea-c1b9-4076-99be-b109b517bd6b)

This is extremely important, and the only reason to use schema, however Geany is showing me only the type and the scope, which is empty, and I don't know what is this for. I believe having these info is essential to use schema with a LSP, so I recommend to add them as soon as possible.

Thanks... 🙏

**Reply by Cicorione** [reply]
~For unknown reasons Geany started to crash when I invoke/enable the XML server:~

Never mind VSCode uninstalled the extensions where I was taking the server... 🤦‍♂️

**Reply by Cicorione** [reply]
I used this other method:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="/home/freezr/git/publisher/schema/layoutschema-en.xsd" type="application/xml" schematypens="http://http://www.w3.org/2001/XMLSchema"?>
<?xml-model href="/home/freezr/git/publisher/schema/layoutschema-en.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
```

And somehow something does, however is doing it messy, it looks like it works only on some tags like `<Table>` .
If you intentionally add error it is still unable to catch them. I am not really sure it is actually working, I am going to do a test without LSP.

Without LSP active doesn't shows hints in the `<Table>` tags, so it is recognized but it is not working as expected.

Does it make sense?

Thanks...  🙏

**Reply by techee** [reply]
I think the server isn't configured correctly. Have a look at https://github.com/eclipse/lemminx/blob/main/docs/Configuration.md. In the server settings you can set
```ini
initialization_options_file=/home/some_user/init_options.json
```
and set the contents to what the doc says, i.e. something like
```json
"settings": {
  "xml": {
    "trace": {
      "server": "verbose"
    },
    "catalogs": [],
    "logs": {
      "client": true,
      "file": "/home/.../lsp4xml.log"
    },
    "format": {
      "splitAttributes": true,
      "joinCDATALines": false,
      "joinContentLines": false,
      "joinCommentLines": false,
      "spaceBeforeEmptyCloseTag": false,
      "enabled": true
    }
  }
}  
```
Don't forget to add

https://github.com/eclipse/lemminx/blob/main/docs/Configuration.md#validation

and also check

https://github.com/eclipse/lemminx/blob/main/docs/Features.md#validation

(I don't know exactly what "with extension (XML Schema, DTD validation)" there means - whether that you should install some extra extension or not). Also check

https://github.com/eclipse/lemminx/blob/main/docs/Configuration.md#file-associations

- this is probably how you specify XML schema.

**Comment by Cicorione** [comment]
I get it, it is enabling schema through `RelaxNG`:

```xml
<?xml-model href="/home/cicorione/git/publisher/schema/layoutschema-en.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
```

However is doing it wrong... 🤔

**Reply by Cicorione** [reply]
Documentation says to use `Method 3`

https://github.com/eclipse/lemminx/blob/main/docs/Features.md

🤔

**Reply by techee** [reply]
>> For unknown reasons Geany started to crash when I invoke/enable the XML server:

> Never mind VSCode uninstalled the extensions where I was taking the server... 🤦‍♂️

But even in this case it shouldn't crash. If you are able to reproduce the crash, I'd be interested in getting a backtrace to know where the crash happened:
1. Run `gdb geany`
2. Type `r` and press enter
3. Try to reproduce the crash
4. When the crash happens, type `bt` to get the backtrace
5. Post the result here
6. Quit gdb by `q`

**Reply by Cicorione** [reply]
Here my Linux box:
```
OS: Devuan 5 daedalus
Kernel: aarch64 Linux 6.1.54-10001-gd517e36a604d
```

#### Steps to reproduce the bug:
1. Open Geany
2. Open a project
3. Open the user conf to modify the server path to a wrong one
4. Restart the server
5. Open the tab with the XML file
6. Crash

#### Backtrace:
```sh
Thread 69 "pool-geany" received signal SIGPIPE, Broken pipe.
[Switching to Thread 0xffffefffeee0 (LWP 5167)]
__GI___libc_write (nbytes=1555, buf=0xaaaaaba1f400, 
    fd=12) at ../sysdeps/unix/sysv/linux/write.c:26
26	../sysdeps/unix/sysv/linux/write.c: No such file or directory.
(gdb) bt
#0  __GI___libc_write
    (nbytes=1555, buf=0xaaaaaba1f400, fd=12)
    at ../sysdeps/unix/sysv/linux/write.c:26
#1  __GI___libc_write
    (fd=12, buf=0xaaaaaba1f400, nbytes=1555)
    at ../sysdeps/unix/sysv/linux/write.c:24
#2  0x0000fffff6cb9218 in  ()
    at /lib/aarch64-linux-gnu/libgio-2.0.so.0
#3  0x0000fffff6c6b2fc in g_output_stream_write ()
    at /lib/aarch64-linux-gnu/libgio-2.0.so.0
#4  0x0000fffff6c6b2fc in g_output_stream_write ()
    at /lib/aarch64-linux-gnu/libgio-2.0.so.0
#5  0x0000fffff6c6b4a4 in g_output_stream_write_all ()
    at /lib/aarch64-linux-gnu/libgio-2.0.so.0
#6  0x0000fffff6c6b81c in  ()
    at /lib/aarch64-linux-gnu/libgio-2.0.so.0
#7  0x0000fffff6c8e9f0 in  ()
    at /lib/aarch64-linux-gnu/libgio-2.0.so.0
#8  0x0000fffff6a4342c in  ()
    at /lib/aarch64-linux-gnu/libglib-2.0.so.0
#9  0x0000fffff6a42854 in  ()
    at /lib/aarch64-linux-gnu/libglib-2.0.so.0
#10 0x0000fffff7a7ee58 in start_thread
    (arg=0xffffffffd377) at ./nptl/pthread_create.c:442
#11 0x0000fffff7ae7f9c in thread_start ()
    at ../sysdeps/unix/sysv/linux/aarch64/clone.S:79
```

**Reply by techee** [reply]
Ah, ok, there seem to be multiple pages od the backtrace (when you press enter, other pages will start appearing. I'd need this from the point where geany or geany-lsp sources start appearing.

**Reply by Cicorione** [reply]
Hi @techee 

This is what happens the exact moment I trigger the bug and I digit `bt`:

```sh
Thread 66 "pool-geany" received signal SIGPIPE, Broken pipe.
[Switching to Thread 0xffffef7eeee0 (LWP 4338)]
__GI___libc_write (nbytes=1555, buf=0xaaaaabc8fc00, fd=12)
    at ../sysdeps/unix/sysv/linux/write.c:26
26	../sysdeps/unix/sysv/linux/write.c: No such file or directory.
(gdb) bt
#0  __GI___libc_write (nbytes=1555, buf=0xaaaaabc8fc00, fd=12)
    at ../sysdeps/unix/sysv/linux/write.c:26
#1  __GI___libc_write (fd=12, buf=0xaaaaabc8fc00, nbytes=1555)
    at ../sysdeps/unix/sysv/linux/write.c:24
#2  0x0000fffff6cb9218 in  ()
    at /lib/aarch64-linux-gnu/libgio-2.0.so.0
#3  0x0000fffff6c6b2fc in g_output_stream_write ()
    at /lib/aarch64-linux-gnu/libgio-2.0.so.0
#4  0x0000fffff6c6b2fc in g_output_stream_write ()
    at /lib/aarch64-linux-gnu/libgio-2.0.so.0
#5  0x0000fffff6c6b4a4 in g_output_stream_write_all ()
    at /lib/aarch64-linux-gnu/libgio-2.0.so.0
#6  0x0000fffff6c6b81c in  ()
```

Thanks... 🙏

**Reply by techee** [reply]
I don't seem to be able to reproduce this issue here. Moreover, `SIGPIPE` should be ignored by Geany itself so this shouldn't happen. Do you still experience this problem when you try the steps you described above?

**Comment by techee** [comment]
All right, I installed vscode with the plugin to investigate more and found this issue:

https://github.com/eclipse/lemminx/issues/1674

I've added a workaround to the plugin so if you git pull the latest version of the plugin and recompile it, it should start working much better (well, document syncing didn't work at all because of this problem).

Second thing to note - when you start playing with the configuration, enclose the whole configuration from https://github.com/techee/geany-lsp/discussions/41#discussioncomment-10479170 into `{}` so the result looks like this:
```json
{
  "settings": {
    "xml": {
      "trace": {
        "server": "verbose"
      },
      "catalogs": [],
      "logs": {
        "client": true,
        "file": "/home/cicorione/lsp4xml.log"
      },
      "format": {
        "splitAttributes": true,
        "joinCDATALines": false,
        "joinContentLines": false,
        "joinCommentLines": false,
        "spaceBeforeEmptyCloseTag": false,
        "enabled": true
      }
    }
  }
}
```

I didn't investigate more but this should bring you to the right track.

**Reply by Cicorione** [reply]
I improved a bit the situation adding also the catalog in the `catalog` section, but I am not sure I am enabling the XML schema the right way.

**Comment by techee** [comment]
Seems to work on my machine - I took some sample from www.w3schools.com and created `note.xsd` with
```xml
<?xml version="1.0"?>
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema"
targetNamespace="https://www.w3schools.com"
xmlns="https://www.w3schools.com"
elementFormDefault="qualified">

<xs:element name="note">
  <xs:complexType>
    <xs:sequence>
      <xs:element name="to" type="xs:string"/>
      <xs:element name="from" type="xs:string"/>
      <xs:element name="heading" type="xs:string"/>
      <xs:element name="body" type="xs:string"/>
    </xs:sequence>
  </xs:complexType>
</xs:element>

</xs:schema>
```

Then I took
```xml
<?xml version="1.0"?>

<note xmlns="https://www.w3schools.com"
xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
xsi:schemaLocation="https://www.w3schools.com note.xsd">

<to>Tove</to>
<from>Jani</from>
<heading>Reminder</heading>
<body>Don't forget me this weekend!</body>
</note>
```

and modified it to contain invalid xml and this is what I get:

<img width="559" alt="Screenshot 2024-08-29 at 23 36 03" src="https://github.com/user-attachments/assets/c15e5909-e5ef-40f0-9ed2-d18c3f499618">

**Reply by Cicorione** [reply]
That's good, meanwhile I can't get the schema I need able to work:

![geany-lsp](https://github.com/user-attachments/assets/5498c50d-7e29-4f2d-acb6-c59752493595)

This file `layoutschema-en.rng` is exactly in the path the pop-up is telling me that it does not exist... 😵‍💫

I am also passing the `catalog` in the `json` file, it should help process of recognize the schema and it looks like it is properly loaded:

```sh
10:14:41: Info: Aug 30, 2024 10:14:41 org.eclipse.lemminx.extensions.contentmodel.uriresolver.XMLCatalogResolverExtension setCatalogs()
Message: Adding XML catalog '/home/cicorione/git/publisher/schema/catalog-schema-en.xml' with expand system id 'file:///home/cicorione/git/publisher/schema/catalog-schema-en.xml' and root URI 'file:///home/cicorione/sp/agl/'.
```

I am sorry I am very dumb at these stuff... 😭

**Comment by Cicorione** [comment]
I do believe I made this working for my use case, you need the catalog assigned in the `json` file and then you need this header:

```
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="/home/freezr/git/publisher/schema/layoutschema-en.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
```

That I am already using with the `scholarly XML` for VSCode. What is not working quite right is auto-completion, perhaps it may depends by the `json` file and I need to specify that feature... 🤷‍♂️

What is **definitely** is a message in the status bar that says the XML is *valid* or *invalid* as in this example:

![vscode-bar](https://github.com/user-attachments/assets/91c4c0ea-c1b9-4076-99be-b109b517bd6b)

Do I need to open a bug request to you or to the general Geany project?

Thanks... 🙏

**Reply by techee** [reply]
> What is definitely is a message in the status bar that says the XML is valid or invalid as in this example:

What I suspect this does is that it checks the number of diagnostic messages for the file (those underlined in the document) and when there are none, it shows "XML is well formed". I want the plugin to be language-agnostic so I won't add such a hard-coded message. What I could try to do (I'm not sure if Geany's API allows me to write to the status bar, I'll have to use some hack) is that I'll add something like
```
Issues: 0
```
to show the number of diagnostic messages for the current file - when it shows 0, it will mean the document is validated. Would that be sufficient?

**Reply by Cicorione** [reply]
I do understand your point and I am agreed with you. I wonder if "xml invalid" can be outputed on the status tab, this would be a win-win combination! 👍

**Reply by techee** [reply]
I added this

<img width="232" alt="Screenshot 2024-09-01 at 20 31 47" src="https://github.com/user-attachments/assets/6f50dc33-1321-42f4-a099-e175e9dcc202">

to the right corner of the status bar - when there are some issues, it shows how may of them are. Also, when the LSP server isn't running or isn't configured for the file, this isn't displayed at all so when you see `issues: 0`, you can be sure there are no problems with the current file.

> I do understand your point and I am agreed with you. I wonder if "xml invalid" can be outputed on the status tab, this would be a win-win combination! 👍

I really don't want to do anything specific to concrete language servers and `issues: 0` is wording that is valid for any language.

**Comment by Cicorione** [comment]
When I use these settings the server won't work:
```json
{
  "settings": {
    "xml": {
      "trace": {
        "server": "verbose"
      },
      "catalogs": ["/home/cicorione/git/publisher/schema/catalog-schema-en.xml"],
      "logs": {
        "client": true,
        "file": "/home/freezr/lsp4xml.log"
      },
      "format": {
        "splitAttributes": true,
        "joinCDATALines": false,
        "joinContentLines": false,
        "joinCommentLines": false,
        "spaceBeforeEmptyCloseTag": false,
        "enabled": true
      },
      "completion": {
	"autoCloseTags": true
      },
	"validation": {
	 "noGrammar": "hint",
	 "enabled": true,
	 "schema": true
     }
   }
  }
}
```

The documentation states:

> Note: all completion settings are by default, true.

However it is not working the way I expect...

😩

**Reply by techee** [reply]
> However it is not working the way I expect...

What exactly is/is not happening? Does the log file say anything useful?

**Reply by Cicorione** [reply]
Log doesn't say anything.

What is not working as I expected, based on the VScode extensions experience, is:

1. Autocompletion tag
2. Only the tags allowed are suggested
3. Attributes are suggested after the tag.

None of that is happening. Tags do not complete themselves, I get wrong tags suggestions, and attributes are mixed together with the tags.

This makes the use of the schema really ineffective...

😔

**Reply by techee** [reply]
Alright, I'm not seeing this but I think we need to have a look at the same xml and the same xsd - can you share something with me where you are able to reproduce the problem?

**Reply by Cicorione** [reply]
Hi @techee 

Speedata Publisher Schema are available here:

https://github.com/speedata/publisher/tree/develop/schema

This is an example how ScholarlyXML VSCode plugin works:

![vscode](https://github.com/user-attachments/assets/e0362c0b-f77d-40ff-8c56-1897bee8fb3b)

I couldn't do a screencast of Geany because I was unable to run the server, which is another version, specifically for `x86`, that doesn't require java to work. I don't know if this is another bug or a pebcak, the `lemminx-binary` is available here: 
 
https://download.jboss.org/jbosstools/vscode/stable/lemminx-binary/0.19.1-541/

Last but not least XML code that you can use as sample to work with:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="/usr/share/speedata-publisher/schema/layoutschema-en.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>

<Layout xmlns="urn:speedata.de:2009/publisher/en"
  xmlns:sd="urn:speedata:2009/publisher/functions/en">

  <Options
    mainlanguage="English (Great Britain)"/>
  <Trace grid="no" gridallocation="no"/>


  <LoadFontfile name="EB08-Regular" filename="EBGaramond08-Regular.ttf" marginprotrusion="60" />
  <LoadFontfile name="EB12-Regular" filename="EBGaramond12-Regular.ttf" />

  <DefineFontfamily name="Title" fontsize="14pt" leading="16pt">
    <Regular fontface="EB12-Regular"/>
  </DefineFontfamily>
  <DefineFontfamily name="text" fontsize="10pt" leading="12pt">
    <Regular fontface="EB08-Regular"/>
  </DefineFontfamily>

  <DefineTextformat name="text" orphan="yes"/>
  <DefineTextformat name="Title" margin-top="3pt" orphan="yes"/>

  <SetGrid width="5mm" height="12pt"/>
  <Pageformat width="125mm" height="145mm"/>

  <Pagetype name="text" test="true()">
    <Margin left="1cm" right="1cm" top="1cm" bottom="1cm"/>
    <PositioningArea name="text">
      <PositioningFrame width="10" height="30" row="1" column="1"/>
      <PositioningFrame width="10" height="30" row="1" column="12"/>
    </PositioningArea>
  </Pagetype>

  <Record element="data">
    <PlaceObject  row="1" column="2" allocate="yes">
      <Image file="sample.pdf" width="20" />
    </PlaceObject>
    <PlaceObject row="9" column="12" valign="bottom">
      <Image file="tree.pdf" width="10"/>
    </PlaceObject>
    <ForAll select="section">
      <Output area="text">
        <Text>
          <Paragraph textformat="Title" fontfamily="Title">
            <Value select="@title"/>
          </Paragraph>
        </Text>
      </Output>
      <Output area="text">
        <Text>
          <Paragraph fontfamily="text">
            <Value select="string(.)"/>
          </Paragraph>
        </Text>
      </Output>
    </ForAll>
  </Record>
</Layout>
```

**Comment by techee** [comment]
> This is an example how ScholarlyXML VSCode plugin works:

It's weird because this "works" for me (with some problems which I'll investigate). I just downloaded the schema you pointed me to, updated `<?xml-model>` inside the xml document to point to the downloaded schema and validation plus autocompletion started working.

For reference, this is the LSP plugin settings I use:
```ini
[XML]
cmd=java -jar /home/parallels/.vscode/extensions/redhat.vscode-xml-0.27.1/server/org.eclipse.lemminx-0.28.0-uber.jar
initialization_options_file=/home/parallels/init_options.json
```

and this is the contents of `init_options.json`:
```json
{
  "settings": {
    "xml": {
      "trace": {
        "server": "verbose"
      },
      "catalogs": [],
      "logs": {
        "client": true,
        "file": "/home/parallels/lsp4xml.log"
      },
      "format": {
        "splitAttributes": true,
        "joinCDATALines": false,
        "joinContentLines": false,
        "joinCommentLines": false,
        "spaceBeforeEmptyCloseTag": false,
        "enabled": true
      }
    }
  }
}
```

Now to what "works" means:
1. Entries in the autocompletion list don't seem to be sorted as you type when prefixed by `<` - I have to check what's going on
2. After entering the autocompletion suggestion, the cursor gets placed behind the close tag and not between the beginning tag and the ending tag - I have to investigate what's going on too

>  I don't know if this is another bug or a pebcak, the lemminx-binary is available here:

I'm on an ARM64 macOS machine and cannot investigate this easily. First I'd suggest to run the binary from the command-line to see if you can execute it. Also, when started, it should just block until you press Ctrl+C. Then, check what the Status window of Geany says and if this server is getting started or not.

**Reply by Cicorione** [reply]
The server behavior should improve if you also add the catalog:

```
[...]
"catalogs": ["/your/path/catalog-schema-en.xml"],
[...]
```

Regarding this:

>  First I'd suggest to run the binary from the command-line to see if you can execute it. Also, when started, it should just block until you press Ctrl+C. Then, check what the Status window of Geany says and if this server is getting started or not.

I tested it the way you wrote but unfortunately the status message doesn't say anything thus I guess it is not initialized at all... I continue to test it to spout out any `pebcak`.

Lastly, thank you very much for taking care of this, having Geany fully proficient with XML schema is a big opportunity for the opensource community.

**Reply by Cicorione** [reply]
I quad checked each path related to the server and are corrects, it would be useful if the status message can say if it was able to connect or not to the server and be more verbose about what is going wrong... 🙏

Thanks... 🙏

**Reply by techee** [reply]
I've added information about server process start to the status window - the other information was already present. You should see something like:

<img width="836" alt="Screenshot 2024-09-01 at 20 15 46" src="https://github.com/user-attachments/assets/880edc1c-5986-46e8-ab73-3c1bb2d9446b">

The last message (truncated in the screenshot) says the server was initialized which means the initial handshake between the client and the server was successful (means the connection was successful and the client and the server exchanged the information about their capabilities). You can see what the server sent to the client when you go to Tools->LSP Client->Server initialize responses.

> Lastly, thank you very much for taking care of this, having Geany fully proficient with XML schema is a big opportunity for the opensource community.

I'm actually really happy that someone tries it for something else than normal programming languages which I tested a bit. The requirements for XML are somewhat different and better to find them now than after the release.

**Comment by Cicorione** [comment]
I found the pebcak, there was an issue in the `json` config file, also `lemminx-linux` requires to define a catalog in order to work; however it still doesn't autocomplete some tag and still continues to suggest the wrong tags, for instance in this screencast it is not able to find the tag `PlaceObject` which is supposed to be valid only in the `Record` section. Afterwards I noticed the attributes hints are correct although it shifts the cursor outside the quotation marks as you already pointed out.

Rewatching the gif I realized that after putting the right tag (`PlaceObject`) all the other hints are more correct including the children tags allowed (but still some attributes in there), I don't know why it messes with the parent tag though.

![geany-lsp](https://github.com/user-attachments/assets/10d1d24a-83fe-43d5-a507-7b023e467e48)

**Reply by techee** [reply]
> however it still doesn't autocomplete some tag and still continues to suggest the wrong tags, for instance in this screencast it is not able to find the tag PlaceObject which is supposed to be valid only in the Record section.

This was the autocompletion sorting issue I mentioned. It should be fixed now but you need to use this config option in the `[XML]` section:
```ini
autocomplete_use_label=true
```

The cursor placement problem after performing autocompletion is caused by the fact that the plugin currently doesn't support autocompletion snippets - I'll try to add at least some basic support for them.

**Reply by Cicorione** [reply]
Hi, the XML behavior is getting better!
However on the earlier commits pressing space bar lets the attributes being pop out, now to get the hints pressing any other letter is fine but the space bar.

![geany-lsp-02](https://github.com/user-attachments/assets/c1d29a0c-3ae3-4e15-85aa-1b6a91fe7046)

There are still some quirks but it is a huge leap forward, for me, and in general nonetheless.

👏👏👏

**Reply by Cicorione** [reply]
Hi @techee

this is a silly question... 😅
On the taskbar after `scope:` is written `unknown`; is there way a to display `schema` instead... 🤷‍♂️

**Comment by techee** [comment]
> The cursor placement problem after performing autocompletion is caused by the fact that the plugin currently doesn't support autocompletion snippets - I'll try to add at least some basic support for them.

This should be fixed now when you add this settings under the [XML] section:
```ini
autocomplete_use_snippets=true
```

You might also want to add
```ini
diagnostics_statusbar_severity=4
```
to show issues in the status bar with lower severity - e.g. when schema link is missing in the XML document.

> However on the earlier commits pressing space bar lets the attributes being pop out, now to get the hints pressing any other letter is fine but the space bar.

I assume by "hints" you mean the autocompletion window, right? I think in the previous commits autocompletion just didn't work right - there's no space in the server's response for autocompletion trigger characters so the autocompletion window shouldn't appear after space.

> There are still some quirks but it is a huge leap forward, for me, and in general nonetheless.

If there are some problems, please let me know. Maybe test first in vscode though - some problems may be specific to the lemminx server and should possibly be reported there.

> On the taskbar after scope: is written unknown; is there way a to display schema instead... 🤷‍♂️

This comes from Geany, not the plugin, see https://www.geany.org/manual/current/index.html#statusbar-templates so this cannot be controlled by the plugin.

The LSP plugin doesn't know anything about XML, schemas, etc. These are all handled by the server and I don't plan to do any language-specific things as I want the plugin to work more or less identically for all servers and languages.

**Reply by Cicorione** [reply]
@techee 

Now it is working like a charm, thanks! 🙏
The only thing I am not very sure about is the way the attributes are invoked, now pressing any key but the space bar shows a dropdown menu, is this behavior intentionally made by you or it is just a coincidence? 

I believe this feeling belongs to my habits and it is not an issue by any means!

Also I was able to modify the taskbar removing that voice scope... 😁

I look forward to integrate my workflow with Geany, now I can work even on my board, that is amazing!

Thank you very much for your commitment and to make XML schema on Geany a reality! 🙏

**Comment by techee** [comment]
> The only thing I am not very sure about is the way the attributes are invoked, now pressing any key but the space bar shows a dropdown menu, is this behavior intentionally made by you or it is just a coincidence?

I'm not quite sure I understand what you are talking about - could you post a screenshot?

> Thank you very much for your commitment and to make XML schema on Geany a reality! 🙏

Good to hear the plugin is useful for someone!

**Reply by Cicorione** [reply]
Never mind, I can't post a screenshot about that.
It works and I am happy! 🙏

**Comment by techee** [comment, marked as answer]
Alright, so just to summarize this discussion for anyone wanting to use the lemminx server as the "right answer" is scattered all over the place. The configuration should look something like this:

```ini
[XML]
cmd=lemminx
initialization_options_file=/home/my_home/init_options.json
autocomplete_use_label=true
autocomplete_use_snippets=true
```
when using the `lemminx` binary which is on PATH (or alternatively using the full path), or replace `cmd` with `cmd=java -jar lemminx.jar` when using jar (or whatever the `jar` file is named).

The `init_options.json` file may contain something like this:
```json
{
  "settings": {
    "xml": {
      "trace": {
        "server": "verbose"
      },
      "catalogs": [],
      "logs": {
        "client": true,
        "file": "/home/my_home/lsp4xml.log"
      },
      "format": {
        "splitAttributes": true,
        "joinCDATALines": false,
        "joinContentLines": false,
        "joinCommentLines": false,
        "spaceBeforeEmptyCloseTag": false,
        "enabled": true
      }
    }
  }
}
```

For more information regarding the configuration, see https://github.com/eclipse/lemminx/blob/main/docs/Configuration.md
